File: src/Linkedlist/DoublyLinkedList.java
package Linkedlist;

public class DoublyLinkedList {
    private Node head ;
    public void insertFirst(int value){
        Node node = new Node(value);
        node.next = head;
        node.prev = null;
        if(head!=null) {
            head.prev = node;
        }
        head = node;
    }
    public void insertLast(int value){
        Node node = new Node(value);
        if(head==null){
            insertFirst(value);
            return;
        }
        Node temp = head;
        while(temp.next!=null){
            temp = temp.next;
        }
        node.prev = temp;
        temp.next = node;
        node.next = null;
    }
    public void insert(int index,int value){
        Node node = new Node(value);
        if(index==0){
            insertFirst(value);
            return;
        }
        Node temp = head;
        if(head==null){
            insertFirst(value);
            return;
        }
        int count = 1;
        while(temp.next!=null){
            if(count == index){
                break;
            }
            temp = temp.next;
            count+=1;
        }
        if(temp.next==null){
            insertLast(value);
            return;
        }
        node.next = temp.next;
        node.prev = temp;
        temp.next = node;
        node.next.prev = node;
    }
    public void insertAfter(int after,int value){
        Node beforeNode = find(after);
        Node node = new Node(value);
        node.next = beforeNode.next;
        node.prev = beforeNode;
        beforeNode.next = node;
        if(node.next!=null) {
            node.next.prev = node;
        }
    }
    public Node find(int value){
        Node temp = head;
        while(temp!=null){
            if(temp.value==value){
                return temp;
            }
            temp = temp.next;
        }
        return null;
    }
    public void display(){
        Node node = head;
        Node last = null;
        while(node!=null){
            System.out.print(node.value+"->");
            last = node;
            node = node.next;
        }
        System.out.println("END");
        while(last!=null){
            System.out.print(last.value+"->");
            last = last.prev;
        }
        System.out.println("START");
    }
    private class Node{
        private int value;
        private Node prev;
        private Node next;
        public Node(int value){
            this.value = value;
        }
        public Node(int value,Node prev,Node next){
            this.value = value;
            this.prev = prev;
            this.next = next;
        }
    }
}

File: pythonFiles/LinkedList/DoublyLinkedList.py
class DoublyLinkedList:
    class Node:
        def __init__(self,data=None,prev=None,next=None):
            self._data = data
            self._prev = prev
            self._next = next
    def __init__(self):
        self._head = None
        self._tail = None
        self._size = 0
    def insertFirst(self,value):
        node = self.Node(value)
        if self._head == None:
            self._head = node
            self._tail = self._head
        else:
            node._next = self._head
            self._head._prev = node
            node._prev = None
            self._head = node
        self._size += 1
        return self._size
    def insertLast(self,value):
        if self._head == None:
            return self.insertFirst(value)
        node = self.Node(value)
        self._tail._next = node
        node._prev = self._tail
        self._tail = node
        self._size += 1
        return self._size
    def insert(self,index,value):
        if(index==0):
            return self.insertFirst(value)
        if(index==self._size):
            return self.insertLast(value)
        prev = self.find(index-1)
        node = self.Node(value)
        node._next = prev._next
        prev._next = node
        node._prev = prev
        node._next._prev = node
        self._size += 1
        return self._size
    def find(self,index):
        node = self._head
        for i in range(index):
            node = node._next
        return node
    def displayLinkedList(self):
        node = self._head
        while(node):
            print(str(node._data)+"->",end="")
            node = node._next
        print("None")
    def displayReverse(self):
        node = self._tail
        while(node):
            print(str(node._data)+"->",end="")
            node = node._prev
        print("None")
dll = DoublyLinkedList()
print(dll.insertLast(0))
print(dll.insert(0,100))
dll.displayLinkedList()
print(dll.insertFirst(1))
print(dll.insert(1,1000))
dll.displayLinkedList()
print(dll.insert(3,3000))
dll.displayLinkedList()
print(dll.insertFirst(2))
dll.displayLinkedList()
dll.displayReverse()